### ISSUE TITLE
error: command 'g++' failed with exit status 1

### ISSUE THREAD

**Issue by dimitri320** [madhawav/YOLO3-4-Py#106]
I'm running on GCP Debian 9, with Python 3.6 using Anaconda to manage my environments.

I've tried to run

`pip install yolo34py-gpu`

But getting the following error:

> ERROR: Command errored out with exit status 1:
>      command: /home/dimitri/anaconda3/envs/yolov3_python3.6/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-91mpp087/yolo34py-gpu/setup.py'"'"'; __file__='"'
> "'/tmp/pip-install-91mpp087/yolo34py-gpu/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__,
>  '"'"'exec'"'"'))' install --record /tmp/pip-record-301du_vu/install-record.txt --single-version-externally-managed --compile
>          cwd: /tmp/pip-install-91mpp087/yolo34py-gpu/
>     Complete output (59 lines):
>     INFO:root:Selected Darknet Branch: yolo34py-intergration-nocv from Darknet Fork 'https://github.com/madhawav/darknet/'
>     /home/dimitri/anaconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/setuptools/dist.py:476: UserWarning: Normalizing '0.1.rc12' to '0.1rc12'
>       normalized_version,
>     running install
>     running build
>     INFO:root:Temp Path: /tmp/darknet
>     INFO:root:Downloading darknet
>     INFO:root:Extracting darknet
>     INFO:root:Building darknet
>     ./src/gemm.c: In function ‘time_gpu’:
>     ./src/gemm.c:232:9: warning: ‘cudaThreadSynchronize’ is deprecated [-Wdeprecated-declarations]
>              cudaThreadSynchronize();
>              ^~~~~~~~~~~~~~~~~~~~~
>     In file included from /usr/local/cuda/include/cuda_runtime.h:96:0,
>                      from include/darknet.h:14,
>                      from ./src/utils.h:5,
>                      from ./src/gemm.c:2:
>     /usr/local/cuda/include/cuda_runtime_api.h:947:57: note: declared here
>      extern __CUDA_DEPRECATED __host__ cudaError_t CUDARTAPI cudaThreadSynchronize(void);
>                                                              ^~~~~~~~~~~~~~~~~~~~~
>     /usr/bin/ld: cannot find -lcuda
>     collect2: error: ld returned 1 exit status
>     make: *** [libdarknet.so] Error 1
>     running build_py
>     creating build
>     creating build/lib.linux-x86_64-3.6
>     creating build/lib.linux-x86_64-3.6/__libdarknet
>     copying __libdarknet/__init__.py -> build/lib.linux-x86_64-3.6/__libdarknet
>     running egg_info
>     writing yolo34py_gpu.egg-info/PKG-INFO
>     writing dependency_links to yolo34py_gpu.egg-info/dependency_links.txt
>     writing requirements to yolo34py_gpu.egg-info/requires.txt
>     writing top-level names to yolo34py_gpu.egg-info/top_level.txt
>     reading manifest file 'yolo34py_gpu.egg-info/SOURCES.txt'
>     reading manifest template 'MANIFEST.in'
>     warning: no files found matching '__libdarknet/libdarknet.so'
>     warning: no files found matching 'pydarknet.cpp'
>     writing manifest file 'yolo34py_gpu.egg-info/SOURCES.txt'
>     running build_ext
>     cythoning pydarknet.pyx to pydarknet.cpp
>     /home/dimitri/anaconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will
>  change in a later release! File: /tmp/pip-install-91mpp087/yolo34py-gpu/pydarknet.pxd
>       tree = Parsing.p_module(s, pxd, full_module_name)
>     building 'pydarknet' extension
>     creating build/temp.linux-x86_64-3.6
>     gcc -pthread -B /home/dimitri/anaconda3/envs/yolov3_python3.6/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DUSE_CV=0 -I/home/dimitri/a
> naconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/numpy/core/include -I/tmp/darknet/darknet-yolo34py-intergration-nocv/include -I/tmp/darknet/darknet-yolo34py-intergration-nocv/src -I/home
> /dimitri/anaconda3/envs/yolov3_python3.6/include/python3.6m -c pydarknet.cpp -o build/temp.linux-x86_64-3.6/pydarknet.o -I/usr/include/python3.5m -I/usr/include/x86_64-linux-gnu/python3.5m
>     cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
>     In file included from /home/dimitri/anaconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/numpy/core/include/numpy/ndarraytypes.h:1832:0,
>                      from /home/dimitri/anaconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
>                      from /home/dimitri/anaconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/numpy/core/include/numpy/arrayobject.h:4,
>                      from pydarknet.cpp:612:
>     /home/dimitri/anaconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with 
> " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
>      #warning "Using deprecated NumPy API, disable it with " \
>       ^~~~~~~
>     gcc -pthread -B /home/dimitri/anaconda3/envs/yolov3_python3.6/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DUSE_CV=0 -I/home/dimitri/a
> naconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/numpy/core/include -I/tmp/darknet/darknet-yolo34py-intergration-nocv/include -I/tmp/darknet/darknet-yolo34py-intergration-nocv/src -I/home
> /dimitri/anaconda3/envs/yolov3_python3.6/include/python3.6m -c bridge.cpp -o build/temp.linux-x86_64-3.6/bridge.o -I/usr/include/python3.5m -I/usr/include/x86_64-linux-gnu/python3.5m
>     cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
>     g++ -pthread -shared -B /home/dimitri/anaconda3/envs/yolov3_python3.6/compiler_compat -L/home/dimitri/anaconda3/envs/yolov3_python3.6/lib -Wl,-rpath=/home/dimitri/anaconda3/envs/yolov3_python3.
> 6/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-x86_64-3.6/pydarknet.o build/temp.linux-x86_64-3.6/bridge.o -L. -L./__libdarknet -ldarknet -lm -lpthread -o build/lib.linux-x86_64-3.6/pyda
> rknet.cpython-36m-x86_64-linux-gnu.so -lpython3.5m -Wl,-rpath,/home/dimitri/anaconda3/envs/yolov3_python3.6/lib/python3.6/site-packages/__libdarknet
>     /home/dimitri/anaconda3/envs/yolov3_python3.6/compiler_compat/ld: cannot find -ldarknet
>     collect2: error: ld returned 1 exit status
>     error: command 'g++' failed with exit status 1
>     ----------------------------------------
> ERROR: Command errored out with exit status 1: /home/dimitri/anaconda3/envs/yolov3_python3.6/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-91mpp087/yolo34p
> y-gpu/setup.py'"'"'; __file__='"'"'/tmp/pip-install-91mpp087/yolo34py-gpu/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.clo
> se();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-301du_vu/install-record.txt --single-version-externally-managed --compile Check the logs for full command output
> .

My GCC version is:

`gcc (Debian 6.3.0-18+deb9u1) 6.3.0`

Not sure how to fix this issue. Any advice will be highly appreciated.

**Comment by madhawav** [comment]
Hi,
Do you have cuda toolkit installed in your machine? Which version is it?

**Comment by dimitri320** [comment]
Here's my NVIDIA drive info:

```
  *-display                 
       description: 3D controller
       product: GP100GL [Tesla P100 PCIe 16GB]
       vendor: NVIDIA Corporation
       physical id: 4
       bus info: pci@0000:00:04.0
       version: a1
       width: 64 bits
       clock: 33MHz
       capabilities: msi pm bus_master cap_list
       configuration: driver=nvidia latency=0
       resources: iomemory:80-7f iomemory:c0-bf irq:47 memory:fc000000-fcffffff memory:800000000-bffffffff memory:c
00000000-c01ffffff ioport:c000(size=128)
```


And the CUDA info:

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Nov__3_21:07:56_CDT_2017
Cuda compilation tools, release 9.1, V9.1.85
```

**Comment by Sunwukong00** [comment]
I have same problem,and  how to fix this issue. Any advice will be highly appreciated

**Comment by madhawav** [comment]
Hi @dimitri320 ,
The error says linker is unable to find CUDA installation. If CUDA ToolKit is installed (which seems to be the case since nvcc works), then maybe you do not have CUDA toolkit libraries added to the linker. Hopefully, the following procedure should work.

Check the location of your CUDA installation. Usually, its installed at `/usr/local/cuda/` and then if so, set the LD_LIBRARY_PATH environment variable as follows.
```bash
export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH
```
If Cuda is located at a different path, change the above path accordingly. Then re-run the installation

**Comment by Sunwukong00** [comment]
Hi guys!
Thank your advice.I have checked location of my CUDA installation and environment variable,both of them are ok.When i try to again run `pip install -r requirements.txt`.
But getting the new error:

>Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple Requirement already satisfied: tensorflow-gpu in /home/scw4750/anaconda3/envs/lcx_glow/lib/python3.5/site-packages (from -r
requirements.txt (line 1)) (1.10.0)
Requirement already satisfied: keras==2.2.0 in /home/scw4750/anaconda3/envs/lcx_glow/lib/python3.5/site-packages (from -r requirements.txt (line 3)) (2.2.0)
Requirement already satisfied: pillow==5.2.0 in /home/scw4750/anaconda3/envs/lcx_glow/lib/python3.5/site-packages (from -r requirements.txt (line 4)) (5.2.0)
Requirement already satisfied: toposort==1.5 in /home/scw4750/anaconda3/envs/lcx_glow/lib/python3.5/site-packages (from -r requirements.txt (line 5)) (1.5)
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.tuna.tsinghua.edu.cn', port=443): Read timed out. (read timeout=15)",)': /simple/horovod/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.tuna.tsinghua.edu.cn', port=443): Read timed out. (read timeout=15)",)': /simple/horovod/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.tuna.tsinghua.edu.cn', port=443): Read timed out. (read timeout=15)",)': /simple/horovod/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.tuna.tsinghua.edu.cn', port=443): Read timed out. (read timeout=15)",)': /simple/horovod/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.tuna.tsinghua.edu.cn', port=443): Read timed out. (read timeout=15)",)': /simple/horovod/
ERROR: Could not find a version that satisfies the requirement horovod==0.13.8 (from -r requirements.txt (line 6)) (from versions: none)
ERROR: No matching distribution found for horovod==0.13.8 (from -r requirements.txt (line 6))

**Comment by madhawav** [comment]
@Sunwukong00 This is a net connectivity related issue. Maybe try again later, or try upgrading pip.
